### PR TITLE
Assessments

### LINKED ISSUE #26: Implement Changeset Test for Assessments Context
Right now all of them have no tests. Follow https://github.com/source-academy/cadet/tree/master/test/cadet/course

### PR DESCRIPTION
**Pull request by sreycodes**
Fixes #26 #28

### CODE REVIEW

File: lib/cadet/assessments/problem_type.ex [hunk 1]
```diff
@@ -0,0 +1,5 @@
+import EctoEnum
+
+defenum Cadet.Assessments.ProblemType, :type, [
```
[the remark this answers is not in the document]
Comment (sreycodes) [reply lib/cadet/assessments/problem_type.ex]: Enum must be used for question and answer I think

[the remark this answers is not in the document]
Comment (indocomsoft) [reply lib/cadet/assessments/problem_type.ex]: Fair enough

File: lib/cadet_web/controllers/assessments_controller.ex [hunk 2]
```diff
@@ -0,0 +1,77 @@
+defmodule CadetWeb.AssessmentController do
+  use CadetWeb, :controller
+
+  alias Cadet.Assessments
+
+  @tabs_map %{
```
Comment (indocomsoft) [inline lib/cadet_web/controllers/assessments_controller.ex]: Why do we have tabs?

File: lib/cadet_web/controllers/assessments_controller.ex [hunk 3]
```diff
@@ -0,0 +1,77 @@
+defmodule CadetWeb.AssessmentController do
```
Comment (indocomsoft) [inline lib/cadet_web/controllers/assessments_controller.ex]: Please write swagger specification so I can review your code better :)

File: lib/cadet_web/controllers/question_controller.ex [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+      changeset: changeset
+    })
+  end
+
+  def update(conn, params) do
+    mission_id = params["mission_id"]
+    Assessments.update_question(params["id"], params["question"])
+  end
+
+  def delete(conn, %{"mission_id" => mission_id, "id" => id}) do
```
Comment (indocomsoft) [inline lib/cadet_web/controllers/question_controller.ex]: Unclosed def

**Comment by indocomsoft** [comment]
@sreycodes I think we should separate the controller PR from the model PR -- this PR is already very large.

File: lib/cadet/assessments.ex [hunk 5]
```diff
@@ -0,0 +1,590 @@
...
+        if question.type == "programming" do
+          prepare_programming_question(
+            question,
+            submission
+          )
+        else
+          %{type: :mcq_question}
+        end
+
+      Poison.encode!(
```
Comment (indocomsoft) [inline lib/cadet/assessments.ex]: Why JSON encode here? Encoding should be the task of the controller

File: lib/cadet/assessments.ex [hunk 6]
```diff
@@ -0,0 +1,590 @@
...
+    mission = Repo.get(Mission, id)
+
+    if mission == nil do
+      {:error, :mission_not_found}
+    else
+      prepare_workspace(mission, question_order, student)
+    end
+  end
+
+  def prepare_workspace(mission, question_order, student) do
```
Comment (indocomsoft) [inline lib/cadet/assessments.ex]: Not sure if this belongs here. Might be more appropriate to be put in the controller. @evansb what do you think?

File: lib/cadet/assessments.ex [hunk 7]
```diff
@@ -0,0 +1,590 @@
...
+      t1 == :contest && t2 == :sidequest -> false
+    end
+  end
+
+  def story_name(mission) do
+    type = Atom.to_string(mission.type)
+    String.downcase(type) <> "-" <> mission.name
+  end
+
+  defp create_blank_question(question, type) do
```
Comment (indocomsoft) [inline lib/cadet/assessments.ex]: Why is this function empty?

File: lib/cadet/assessments/Question.ex [hunk 8]
```diff
@@ -0,0 +1,54 @@
+defmodule Cadet.Assessments.Question do
```
Comment (indocomsoft) [inline lib/cadet/assessments/Question.ex]: Please rename the file to all lowercase. Pleaseeeeeeeeeeee

File: lib/cadet/assessments.ex [hunk 9]
```diff
@@ -0,0 +1,590 @@
...
+  end
+
+  defp create_concrete_question(question, type) do
```
Comment (indocomsoft) [inline lib/cadet/assessments.ex]: I don't think we need this since the question type polymorphism is in the JSON. @evansb what's your opinion?

File: lib/cadet/assessments/Question.ex [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+    externals: [],
+    files: []
+  }
+
+  schema "questions" do
+    field(:title, :string)
+    field(:display_order, :integer)
+    field(:weight, :integer)
+    field(:library, :map, default: @default_library)
+    field(:raw_library, :string, virtual: true)
```
Comment (indocomsoft) [inline lib/cadet/assessments/Question.ex]: Please add `put_library` in and add tests for it

File: lib/cadet/assessments/Question.ex [hunk 11]
```diff
@@ -0,0 +1,54 @@
+defmodule Cadet.Assessments.Question do
+  @moduledoc false
```
Comment (indocomsoft) [inline lib/cadet/assessments/Question.ex]: Please add documentation

File: lib/cadet/assessments/answer.ex [hunk 12]
```diff
@@ -0,0 +1,44 @@
+defmodule Cadet.Assessments.Answer do
+  @moduledoc false
```
Comment (indocomsoft) [inline lib/cadet/assessments/answer.ex]: Please add documentation

File: lib/cadet/assessments.ex [hunk 13]
```diff
@@ -0,0 +1,590 @@
...
+    Enum.filter(mission_with_category, &(&1.is_published && Timex.before?(&1.open_at, now)))
+  end
+
+  def missions_due_soon() do
+    now = Timex.now()
+    week_after = Timex.add(now, Duration.from_weeks(1))
+
+    all_missions()
+    |> Enum.filter(
+      &(&1.is_published && Timex.before?(&1.open_at, now) &&
```
Comment (indocomsoft) [inline lib/cadet/assessments.ex]: use `and` instead of `&&` similarly for the rest of the file. Refer to https://github.com/lexmag/elixir-style-guide#boolean-operators

File: lib/cadet/assessments/mission.ex [hunk 14]
```diff
@@ -8,19 +8,32 @@ defmodule Cadet.Assessments.Mission do
...
     field(:cover_picture, Image.Type)
+    has_many(:questions, Question, on_delete: :delete_all)
+    field(:order, :string, default: "")
+    timestamps()
   end
 
-  @required_fields ~w(order category title open_at close_at)a
-  @optional_fields ~w(summary_short summary_long)
+  @required_fields ~w(category title open_at close_at)a
+  @optional_fields ~w(summary_short summary_long is_published max_xp priority)a
```
Comment (indocomsoft) [inline lib/cadet/assessments/mission.ex]: `max_xp` should be required

File: lib/cadet/assessments/mission.ex [hunk 15]
```diff
@@ -8,19 +8,32 @@ defmodule Cadet.Assessments.Mission do
...
   @optional_file_fields ~w(cover_url)
+
+  def changeset(mission, params) do
+    mission
+    |> cast(params, @required_fields ++ @optional_fields)
+    |> validate_required(@required_fields)
+    |> validate_number(:max_xp, greater_than_or_equal_to: 0)
```
Comment (indocomsoft) [inline lib/cadet/assessments/mission.ex]: You need to process the `cover_url`

File: lib/cadet/assessments/mission.ex [hunk 16]
```diff
@@ -8,19 +8,32 @@ defmodule Cadet.Assessments.Mission do
...
+  end
```
Comment (indocomsoft) [inline lib/cadet/assessments/mission.ex]: Also `validate_open_close_date`

File: test/cadet/assessments/answer_test.exs [hunk 17]
```diff
@@ -0,0 +1,26 @@
+defmodule Cadet.Assessments.AnswerTest do
```
Comment (indocomsoft) [inline test/cadet/assessments/answer_test.exs]: Test by putting in json in `raw_answer`

File: test/cadet/assessments/question_test.exs [hunk 18]
```diff
@@ -0,0 +1,39 @@
+defmodule Cadet.Assessments.QuestionTest do
+  use Cadet.ChangesetCase, async: true
```
Comment (indocomsoft) [inline test/cadet/assessments/question_test.exs]: Test the `raw_library` and `raw_question`

File: lib/cadet/assessments/Question.ex [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+  end
+
+  defp put_question(changeset) do
+    change = get_change(changeset, :raw_question)
+
+    if change do
+      json = Poison.decode!(change)
+
+      if json != nil do
+        put_change(changeset, :json, json)
```
Comment (indocomsoft) [inline lib/cadet/assessments/Question.ex]: Why put change in `:json`?

File: test/cadet/assessments/submission_test.exs [hunk 20]
```diff
@@ -0,0 +1,13 @@
...
+
+  alias Cadet.Assessments.Submission
+
+  valid_changesets Submission do
+    %{
+      status: :submitted,
+      submitted_at: Timex.now(),
+      override_xp: 100,
+    }
+  end
```
Comment (indocomsoft) [inline test/cadet/assessments/submission_test.exs]: Add `invalid_changesets`

File: lib/cadet/context_helper.ex [hunk 21]
```diff
@@ -0,0 +1,62 @@
...
+        Repo.update!(
+          change(next_model, %{
+            display_order: model.display_order
+          })
+        )
+      end
+    end)
+  end
+
+  def toggle_field(queryable, id, field) do
```
Comment (indocomsoft) [inline lib/cadet/context_helper.ex]: What do we use this function for?

File: lib/cadet/context_helper.ex [hunk 22]
```diff
@@ -0,0 +1,62 @@
...
+  def put_display_order(changeset, collection) do
+    if Enum.empty?(collection) do
+      change(changeset, %{display_order: 1})
+    else
+      last = Enum.max_by(collection, & &1.display_order)
+      change(changeset, %{display_order: last.display_order + 1})
+    end
+  end
+
+  def update_display_order(queryable, id, offset) do
```
Comment (indocomsoft) [inline lib/cadet/context_helper.ex]: Are we using this function?

File: lib/cadet/assessments/mission.ex [hunk 23]
```diff
@@ -8,19 +8,32 @@ defmodule Cadet.Assessments.Mission do
 
   alias Cadet.Assessments.Category
   alias Cadet.Assessments.Image
+  alias Cadet.Assessments.Question
 
   schema "missions" do
```
Comment (indocomsoft) [inline lib/cadet/assessments/mission.ex]: Need to add mission pdf as well (look up `Upload` model)

File: test/cadet/assessments_test.exs [hunk 24]
```diff
@@ -0,0 +1,251 @@
+defmodule Cadet.AssessmentsTest do
```
Comment (indocomsoft) [inline test/cadet/assessments_test.exs]: Use coveralls to track functions that has not been tested. You can also track locally by running `mix coveralls.html` and opening the generated report.

For example, there's missing coverage for `missions_due_soon/0`, `build_answer/1`, `change_question/2`, and so on.

File: lib/cadet/assessments/mission.ex (hunk 15 above)
Comment (sreycodes) [reply lib/cadet/assessments/mission.ex]: I am not exactly sure how to process this.

File: lib/cadet/model_helper.ex [hunk 25]
```diff
@@ -0,0 +1,20 @@
+defmodule Cadet.ModelHelper do
```
Comment (indocomsoft) [inline lib/cadet/model_helper.ex]: Try to move all the JSON deserialising in the models here for more DRY-ness (Don't Repeat Yourself)

File: test/cadet/assessments_test.exs [hunk 26]
```diff
@@ -0,0 +1,251 @@
...
+  use Cadet.DataCase
```
Comment (indocomsoft) [inline test/cadet/assessments_test.exs]: Change these tests to use factories, and add factories

File: lib/cadet/assessments.ex [hunk 27]
```diff
@@ -3,29 +3,626 @@ defmodule Cadet.Assessments do
...
+  def build_question(params, type) do
+    Question.changeset(%Question{}, Map.merge(%{:type => type}, params))
+  end
+
+  def build_question(params, type) do
+    case type do
+      :programming -> create_programming_question(%{:type => :programming})
+      :multiple_choice -> create_multiple_choice_question(%{:type => :multiple_choice})
+    end
+    change(changeset, %{raw_library: Poison.encode!(changeset.data.library)})
```
Comment (indocomsoft) [inline lib/cadet/assessments.ex]: `changeset` is not defined here?

File: lib/cadet/assessments/question.ex [hunk 28]
```diff
@@ -0,0 +1,74 @@
...
+  @moduledoc """
+  Questions model contains domain logic for questions management 
+  including programming and multiple choice questions.
+  """
+  use Cadet, :model
+
+  alias Cadet.Assessments.Mission
+  alias Cadet.Assessments.ProblemType
+
+  @default_library %{
```
Comment (indocomsoft) [inline lib/cadet/assessments/question.ex]: Library is already included in ProgrammingQuestion, so this can be removed.

File: lib/cadet/assessments/submission.ex [hunk 29]
```diff
@@ -0,0 +1,47 @@
...
+    |> validate_required(@required_fields)
+    |> validate_role(:student, "student")
+    |> validate_role(:grader, "staff")
+  end
+
+  def validate_role(changeset, user, role) do
+    validate_change(changeset, user, fn ^user, user ->
+      case user.role == role do
+        true -> []
+        false -> [{user, "Access Denied"}]
```
Comment (indocomsoft) [inline lib/cadet/assessments/submission.ex]: change error message to `"does not have the role #{role}"`.

Rule about validation error message is that they will be appended to the field name, so, e.g. they should be all lowercase.

In this case, following your code, the error message will be e.g. `Student Access Denied` which does not make sense as compared to `student does not have the role student`

File: lib/cadet/model_helper.ex (hunk 25 above)
Comment (indocomsoft) [reply lib/cadet/model_helper.ex]: Bump

File: test/cadet/assessments/answer_test.exs (hunk 17 above)
Comment (indocomsoft) [reply test/cadet/assessments/answer_test.exs]: Bump

File: test/cadet/assessments/question_types/programming_question_test.exs [hunk 30]
```diff
@@ -24,7 +24,6 @@ defmodule Cadet.Assessments.QuestionTypes.ProgrammingQuestionTest do
       content: "asd",
       solution_template: "asd",
       solution_header: "asd",
-      solution: "asd",
```
Comment (indocomsoft) [inline test/cadet/assessments/question_types/programming_question_test.exs]: Why is this line deleted?

File: test/cadet/assessments_test.exs [hunk 31]
```diff
@@ -0,0 +1,200 @@
...
+  end
+
+  test "delete question" do
+    mission = insert(:mission)
+    question = insert(:question)
+    Assessments.delete_question(question.id)
+    assert Assessments.get_question(question.id) == nil
+  end
+
+  # test "mission and its questions" do
```
Comment (indocomsoft) [inline test/cadet/assessments_test.exs]: Why is this one commented out?

File: test/cadet/assessments_test.exs [hunk 32]
```diff
@@ -0,0 +1,200 @@
...
+    mission = insert(:mission)
+
+    {:ok, question} =
+      Assessments.create_question(
+        %{
+          title: "question",
+          weight: 5,
+          type: :programming,
+          question: %{},
+          raw_question: "{\"content\": \"asd\", \"solution_template\": \"template\",
```
Comment (indocomsoft) [inline test/cadet/assessments_test.exs]: This might look cleaner if you instead do `Poison.encode!(%{content: "asd", solution_template: "template"})`. Similarly for the other `raw_*` fields

File: test/cadet/assessments_test.exs [hunk 33]
```diff
@@ -0,0 +1,200 @@
...
+  test "create mission" do
+    {:ok, mission} =
+      Assessments.create_mission(%{
+        title: "mission",
+        category: :mission,
+        open_at: Timex.now(),
+        close_at: Timex.shift(Timex.now(), days: 7)
+      })
+
+    assert mission.title == "mission"
```
Comment (indocomsoft) [inline test/cadet/assessments_test.exs]: Change all the `assert`s done in this style to pattern matching instead, e.g. `assert mission = %{title: "mission", category: :mission}`

File: test/cadet/assessments_test.exs [hunk 34]
```diff
@@ -0,0 +1,200 @@
...
+    mission = Assessments.get_mission(mission.id)
+    assert mission.title == "changed_mission"
+  end
+
+  test "all missions with category" do
+    mission = insert(:mission, category: :mission)
+    sidequest = insert(:mission, category: :sidequest)
+    contest = insert(:mission, category: :contest)
+    path = insert(:mission, category: :path)
+    assert mission.id in Enum.map(Assessments.all_missions(:mission), fn m -> m.id end)
```
Comment (indocomsoft) [inline test/cadet/assessments_test.exs]: You can simplify this test. Since you're only adding one for each category, you could simply do `assert [mission] = Assessments.all_missions(:mission)`. Similarly for the other `category`s.

File: test/cadet/assessments_test.exs [hunk 35]
```diff
@@ -0,0 +1,200 @@
...
+    assert sidequest.id in Enum.map(Assessments.all_missions(:sidequest), fn m -> m.id end)
+    assert contest.id in Enum.map(Assessments.all_missions(:contest), fn m -> m.id end)
+    assert path.id in Enum.map(Assessments.all_missions(:path), fn m -> m.id end)
+  end
+
+  test "due missions" do
+    mission_before_now = insert(:mission, close_at: Timex.now(), is_published: true)
+
+    mission_in_timerange =
```
Comment (indocomsoft) [inline test/cadet/assessments_test.exs]: `mission_due` is a better name

File: test/cadet/assessments_test.exs [hunk 36]
```diff
@@ -0,0 +1,200 @@
...
+    contest = insert(:mission, category: :contest)
+    path = insert(:mission, category: :path)
+    assert mission.id in Enum.map(Assessments.all_missions(:mission), fn m -> m.id end)
+    assert sidequest.id in Enum.map(Assessments.all_missions(:sidequest), fn m -> m.id end)
+    assert contest.id in Enum.map(Assessments.all_missions(:contest), fn m -> m.id end)
+    assert path.id in Enum.map(Assessments.all_missions(:path), fn m -> m.id end)
+  end
+
+  test "due missions" do
+    mission_before_now = insert(:mission, close_at: Timex.now(), is_published: true)
```
Comment (indocomsoft) [inline test/cadet/assessments_test.exs]: Might be better to shift `Timex.now()` slightly before to like yesterday just to make sure. (I know by the time `missions_due_soon/0` is run, the time would be behind, but I think intent would be clearer this way)

File: test/cadet/assessments_test.exs [hunk 37]
```diff
@@ -0,0 +1,200 @@
...
+      insert(:mission, close_at: Timex.shift(Timex.now(), days: 4), is_published: true)
+
+    mission_far =
+      insert(
+        :mission,
+        close_at: Timex.shift(Timex.now(), weeks: 2),
+        is_published: true
+      )
+
+    result = Enum.map(Assessments.missions_due_soon(), fn m -> m.id end)
```
Comment (indocomsoft) [inline test/cadet/assessments_test.exs]: `missions_due_ids` is a better name

File: test/cadet/assessments_test.exs [hunk 38]
```diff
@@ -0,0 +1,200 @@
...
+        open_at: Timex.now(),
+        close_at: Timex.shift(Timex.now(), days: 7)
+      })
+
+    Assessments.update_mission(mission.id, %{title: "changed_mission"})
+    mission = Assessments.get_mission(mission.id)
+    assert mission.title == "changed_mission"
+  end
+
+  test "all missions with category" do
```
Comment (indocomsoft) [inline test/cadet/assessments_test.exs]: Please split each category into separate tests

File: test/cadet/assessments_test.exs [hunk 39]
```diff
@@ -0,0 +1,200 @@
...
+        is_published: true
+      )
+
+    result = Enum.map(Assessments.missions_due_soon(), fn m -> m.id end)
+    assert mission_in_timerange.id in result
+    refute mission_far.id in result
+  end
+
+  test "update question" do
+    mission = insert(:mission)
```
Comment (indocomsoft) [inline test/cadet/assessments_test.exs]: What is this for?

File: test/cadet/assessments_test.exs [hunk 40]
```diff
@@ -0,0 +1,200 @@
...
+    question = insert(:question)
+    Assessments.update_question(question.id, %{weight: 10})
+    question = Assessments.get_question(question.id)
+    assert question.weight == 10
+  end
+
+  test "delete question" do
+    mission = insert(:mission)
```
Comment (indocomsoft) [inline test/cadet/assessments_test.exs]: What is this for?

File: test/cadet/assessments_test.exs [hunk 41]
```diff
@@ -0,0 +1,200 @@
...
+    mission_far =
+      insert(
+        :mission,
+        close_at: Timex.shift(Timex.now(), weeks: 2),
+        is_published: true
+      )
+
+    result = Enum.map(Assessments.missions_due_soon(), fn m -> m.id end)
+    assert mission_in_timerange.id in result
+    refute mission_far.id in result
```
Comment (indocomsoft) [inline test/cadet/assessments_test.exs]: you need to add `refute mission_before_now.id in result` too

File: lib/cadet/assessments/answer.ex [hunk 42]
```diff
@@ -0,0 +1,47 @@
...
+
+  alias Cadet.Assessments.Mission
+  alias Cadet.Assessments.ProblemType
+
+  schema "answers" do
+    field(:marks, :float, default: 0.0)
+    field(:answer, :map)
+    field(:type, ProblemType)
+    field(:raw_answer, :string, virtual: true)
+    belongs_to(:submission, Submission)
```
[the remark this answers is not in the document]
Comment (indocomsoft) [reply lib/cadet/assessments/answer.ex]: Done

**Comment by indocomsoft** [comment]
@tuesmiddt I feel this should be ready to merge, but since I made the edits, I think it's good that you have a lookthrough. If everything is alright, I'll merge this tomorrow.

File: lib/cadet/assessments/assessments.ex [hunk 43]
```diff
@@ -3,29 +3,132 @@ defmodule Cadet.Assessments do
...
+  end
+
+  def all_assessments(category) do
+    Repo.all(from(a in Assessment, where: a.category == ^category))
+  end
+
+  def all_open_assessments(category) do
+    now = Timex.now()
+
+    assessment_with_category = Repo.all(from(a in Assessment, where: a.category == ^category))
```
Comment (tuesmiddt) [inline lib/cadet/assessments/assessments.ex]: Can use the `all_assessments(category)` method above but I think just leave it for now. In general I think we should combine this and the `Enum.filter` below into one query string so that the DB engine only gives us the results we want instead of fetching too much data into memory and then filtering it. Maybe leave that for a later refactor.

File: lib/cadet/assessments/assessments.ex [hunk 44]
```diff
@@ -3,29 +3,132 @@ defmodule Cadet.Assessments do
...
+  def build_assessment(params) do
+    Assessment.changeset(%Assessment{}, params)
+  end
+
+  def build_question(params) do
+    Question.changeset(%Question{}, params)
+  end
+
+  def create_assessment(params) do
+    changeset = build_assessment(params)
```
Comment (tuesmiddt) [inline lib/cadet/assessments/assessments.ex]: Use pipeline syntax

File: lib/cadet/assessments/assessments.ex [hunk 45]
```diff
@@ -3,29 +3,132 @@ defmodule Cadet.Assessments do
...
+    simple_update(
+      Question,
+      id,
+      using: &Question.changeset/2,
+      params: params
+    )
+  end
+
+  def publish_assessment(id) do
+    assessment = get_assessment(id)
```
Comment (tuesmiddt) [inline lib/cadet/assessments/assessments.ex]: Use pipeline syntax

File: lib/cadet/assessments/assessments.ex [hunk 46]
```diff
@@ -3,29 +3,132 @@ defmodule Cadet.Assessments do
...
+
+  def get_question(id) do
+    Repo.get(Question, id)
+  end
+
+  def get_assessment(id) do
+    Repo.get(Assessment, id)
+  end
+
+  def create_question(params, assessment_id) do
```
Comment (tuesmiddt) [inline lib/cadet/assessments/assessments.ex]: I think can combine this with `create_question_for_assessment`

File: lib/cadet/assessments/assessments.ex (hunk 43 above)
Comment (indocomsoft) [reply lib/cadet/assessments/assessments.ex]: Good point. I'll write it down as a TODO

File: lib/cadet/assessments/assessments.ex (hunk 44 above)
Comment (indocomsoft) [reply lib/cadet/assessments/assessments.ex]: https://github.com/lexmag/elixir-style-guide#needless-pipeline
Basically for single function call, don't use pipeline syntax.

Comment (tuesmiddt) [reply lib/cadet/assessments/assessments.ex]: i was thinking of 
```
params
|> build_assessment()
|> Repo.insert()
```

File: lib/cadet/assessments/assessments.ex (hunk 45 above)
Comment (tuesmiddt) [reply lib/cadet/assessments/assessments.ex]: how about
```
id
|> get_assessment()
|> change(%{is_published: true})
|> Repo.update()
```

File: lib/cadet/assessments/assessments.ex (hunk 44 above)
Comment (indocomsoft) [reply lib/cadet/assessments/assessments.ex]: Ohh makes sense
